language/move-lang/src/parser/syntax.rs

```diff
@@ -1305,37 +1306,72 @@ fn parse_optional_type_args(tokens: &mut Lexer<'_>) -> Result<Option<Vec<Type>>,
     }
 }
 
+fn token_to_ability(token: Tok, content: &str) -> Option<Ability_> {
+    match (token, content) {
+        (Tok::Copy, _) => Some(Ability_::Copy),
+        (Tok::IdentifierValue, Ability_::DROP) => Some(Ability_::Drop),
+        (Tok::IdentifierValue, Ability_::STORE) => Some(Ability_::Store),
+        (Tok::IdentifierValue, Ability_::KEY) => Some(Ability_::Key),
+        _ => None,
+    }
+}
+
+// Parse a type ability
+//      Ability =
+//          <Copy>
+//          | "drop"
+//          | "store"
+//          | "key"
+fn parse_ability(tokens: &mut Lexer) -> Result<Ability, Error> {
+    let loc = current_token_loc(tokens);
+    match token_to_ability(tokens.peek(), tokens.content()) {
+        Some(ability) => {
+            tokens.advance()?;
+            Ok(sp(loc, ability))
+        }
+        None => Err(vec![(
+            loc,
+            format!(
+                "Unexpected type ability modifier, '{}'. Expected one of: 'copy', 'drop', \
+                 'store', or 'key'",
+                tokens.content()
+            ),
+        )]),
+    }
+}
+
 // Parse a type parameter:
 //      TypeParameter =
 //          <Identifier> <Constraint>?
 //      Constraint =
-//          ":" "copyable"
-//          | ":" "resource"
-fn parse_type_parameter(tokens: &mut Lexer<'_>) -> Result<(Name, Kind), Error> {
+//          ":" <Ability> (+ <Ability>)*
+fn parse_type_parameter(tokens: &mut Lexer<'_>) -> Result<(Name, Vec<Ability>), Error> {
     let n = parse_identifier(tokens)?;
 
-    let kind = if match_token(tokens, Tok::Colon)? {
-        let start_loc = tokens.start_loc();
-        let k = match tokens.peek() {
-            Tok::Copyable => Kind_::Affine,
-            Tok::Resource => Kind_::Resource,
-            _ => {
-                let expected = "either 'copyable' or 'resource'";
-                return Err(unexpected_token_error(tokens, expected));
-            }
-        };
-        tokens.advance()?;
-        let end_loc = tokens.previous_end_loc();
-        spanned(tokens.file_name(), start_loc, end_loc, k)
+    let ability_constraints = if match_token(tokens, Tok::Colon)? {
+        parse_list(
+            tokens,
+            |tokens| {
+                if tokens.peek() == Tok::Plus {
+                    tokens.advance()?;
+                    Ok(true)
+                } else {
+                    Ok(!(tokens.peek() == Tok::Greater || tokens.peek() == Tok::Comma))
```

Comment:
I don't understand the check for `Tok::Comma` here. For that matter, it seems like the usual thing we do elsewhere in the parser is to return `Ok(false)` unless we see the list separator (`Tok::Plus` in this case). Or are you intentionally trying to make the `+` separator optional? I don't see that in the grammar comment, and it seems like it would be simpler to require it.

Reply:
Maybe I will pull in my fixes for this from my testing PR into this one. It should make it a bit more clear what I was attempting to do (agreed it is not good as is)